### DISCUSSION TITLE
Layout issue when graph generated using GModel

### DISCUSSION THREAD

**Discussion by sathya-1994** [eclipse-glsp/glsp#454] [General] [answered]
I am using GLSP's vscode extension(https://github.com/eclipse-glsp/glsp-vscode-integration) and language server which extends GLSP server.

I am creating graph using below code/GModel instead of GSON parser. 

**Code:** Using below code to generate GGraph in the org.eclipse.glsp.server.features.core.model.JsonFileGModelLoader class.

			GGraph graph = new GGraphImpl();
                    graph.setId("graph0");
                    graph.setType(DefaultTypes.GRAPH);
                    GNodeImpl gNodeParent = new GNodeImpl();
                    gNodeParent.setId("nodeParent");
                    gNodeParent.setType(DefaultTypes.NODE);
                    gNodeParent.setLayout(GConstants.Layout.VBOX);
                    gNodeParent.setParent(graph);

                    gNodeParent.getLayoutOptions().put(GLayoutOptions.KEY_PADDING_BOTTOM, 7);
			gNodeParent.getLayoutOptions().put(GLayoutOptions.KEY_PADDING_TOP, 7);
			gNodeParent.getLayoutOptions().put(GLayoutOptions.KEY_PADDING_LEFT, 17);
			gNodeParent.getLayoutOptions().put(GLayoutOptions.KEY_PADDING_RIGHT, 17);
			
			GLabelImpl gLabelParent = new GLabelImpl();
			gLabelParent.setId("labelParent");
			gLabelParent.setText("Parent");
			gLabelParent.setType("label:heading");
			gLabelParent.setParent(gNodeParent);
			
			GCompartmentImpl compartmentNode = new GCompartmentImpl();
			compartmentNode.setId("comp1");
			compartmentNode.setType("comp:comp");
			compartmentNode.setLayout(GConstants.Layout.HBOX);
			compartmentNode.setParent(gNodeParent);
			
			GNodeImpl gNodeChild = new GNodeImpl();
			gNodeChild.setParent(compartmentNode);
			gNodeChild.setId("nodeChild");
			gNodeChild.setType(DefaultTypes.NODE);
			gNodeChild.setLayout(GConstants.Layout.VBOX);
			
			GLabelImpl gLabelChild = new GLabelImpl();
			gLabelChild.setId("labelChild");
			gLabelChild.setText("Child");
			gLabelChild.setType("label:heading"); 
                    gLabelChild.setParent(gNodeChild);
			
**Issue:** Initially label "Parent" will be at left corner when we load the graph and when we perform layout operation(alt + L) then "Parent" label will be moved to center of the containing node.

**Initial graph screenshot:**
![image](https://user-images.githubusercontent.com/59145404/140802198-0af08285-ca9a-41ee-8fb5-804cb623e57d.png)

**Graph after layout operation:**
![image](https://user-images.githubusercontent.com/59145404/140802278-94f4fcf7-eacc-4d43-8b33-eae21bdfbb2e.png)

I added vertical alignment, horizontal alignment to the parent and child nodes but still I faced left alignment issue(gNodeChild.getLayoutOptions().put(GLayoutOptions.KEY_V_ALIGN, GConstants.VAlign.CENTER);
			    gNodeChild.getLayoutOptions().put(GLayoutOptions.KEY_H_ALIGN, GConstants.HAlign.CENTER));
			 
**Requirement: Labels should be aligned to center of the containing node when we load the graph.**

**Comment by sathya-1994** [comment]
I debugged little and I found that org.eclipse.glsp.graph.impl.GPointImpl will not be same for nodes which will be calculated at client side(for GLabelImpl and GCompartmentImpl). When we call layout operation, we will get different boundary values (X and Y) for these nodes/labels.

org.eclipse.glsp.server.utils.LayoutUtil.applyBounds(GModelRoot, ComputedBoundsAction, GModelState) here we are copying boundaries generated at client side into graph model.

**Comment by tortmayr** [comment, marked as answer]
Hi, I tested your code sample and to me this looks like an issue with nested GNodes. If I change the implementation of `gNodeChild` to a compartment everything works as expected.
```java
      GCompartmentImpl gNodeChild = new GCompartmentImpl();
      gNodeChild.setParent(compartmentNode);
      gNodeChild.setId("nodeChild");
      gNodeChild.setType("comp:comp");
      gNodeChild.setLayout(GConstants.Layout.VBOX);
```

@CamilleLetavernier  You recently worked on improving the support of structured nodes in GLSP. Are you aware of that issue and/or have an idea why it occurs?

@sathya-1994 FYI: If you want to you can also use your fluent Builder API for creating new GModel  elements e.g:
```java
      GNode gNodeParent = new GNodeBuilder()
         .id("nodeParent")
         .type(DefaultTypes.NODE)
         .layout(GConstants.Layout.VBOX)
         .layoutOptions(new GLayoutOptions()
            .paddingBottom(7d)
            .paddingTop(7d)
            .paddingLeft(7d)
            .paddingRight(7d))
         .build();
```

**Reply by CamilleLetavernier** [reply]
> @CamilleLetavernier You recently worked on improving the support of structured nodes in GLSP. Are you aware of that issue and/or have an idea why it occurs?

Micro-layouts are a bit tricky, because depending on the parent and child layouts (and parameters), sometimes you want the parent to determine the size/position of the child, and sometimes you want to just place the child at 0; 0 and let it determine its own alignment. And some children affect the size of the parent, while the opposite is also true.

I had to override the default Sprotty layout behavior in a few cases to avoid such refresh issues (To make sure client-side layout could get it right the first time, and not just after a client-server layout roundtrip).

I don't think I've tested the case with 2 Labels and a parent with a custom size (That isn't just the size of the largest child-label), but I don't really see why it should be different compared to e.g. a Label and a Compartment, or a single Label (Both of these cases should work as expected with the new VBox layout and resized nodes)

There's really nothing special about Compartments. In the Workflow example, Compartments use vGrab and hGrab options, and a new Layout (FreeForm), but that's just how the workflow example is configured; it's not specific to Compartments.
